WordPress/gutenberg
# Style engine: elements backend support

Tracking issue: https://github.com/WordPress/gutenberg/issues/38167

## What?

1. Updates the style engine to return CSS var values e..g, `color: var(--wp--css--heavenly-blue )` from `'var:preset|color|heavenly-blue'`.
2. Furthermore, if a selector is passed the style engine will return `selector { ...rules }`.
3. Allows registration of block support styles that will ultimately be printed in the header/footer

This enables us to support element styles, which are printed to the page inside a style tag.

For example by passing in the `{ color: { text: 'value' } }` of `{ elements: { link: color: { text: 'value' } } }`
```php
gutenberg_style_engine_generate(
    array( 'color' => '#1a8ff6' ),
    array(
        'selector'                     => ".$class_name a",
        'enqueue_block_support_styles' => true,
    )
);

gutenberg_style_engine_generate(
    array( 'color' => 'var:preset|color|pale-pink' ),
    array(
        'selector'                     => ".$class_name a",
        'css_vars'                     => true,
        'enqueue_block_support_styles' => true,
    )
);
```

After all this, we can print out the backend styles for elements (just links for now).

```html
<style>
	.wp-elements-6b35f975ef7240eac966f882d143c16d a { color: var(--wp--preset--color--pale-pink); }
</style>
<style>
	.wp-elements-6d3be9aa1701f0d822831260687d101e a { color: #1a8ff6; }
</style>
```

## Why?

Elements block supports requires us to generated CSS var values, e.g., `var(--wp--preset--*)`from `var:preset|*` presets.
Furthermore, we need to print the styles into the page via `enqueue_block_support_styles`.


## Testing Instructions

Elements only supports link colors at this stage. So, add a few links with different preset and custom colors applied and make sure they look dandy on the frontend.

<details>

<summary>Example block code</summary>

```html

<p class="has-link-color"><a href="https://test.com">Test</a> paragraph with a link.</p>



<div class="wp-block-group has-link-color">
<p><a href="https://test.com">Test</a> paragraph inside a group with a link.</p>
</div>


```

</details>


<img width="732" alt="Screen Shot 2022-05-13 at 3 01 53 pm" src="https://user-images.githubusercontent.com/6458278/168214475-31572419-7f17-40cb-8be4-91e4028543ff.png">


Run `npm run test-unit-php /var/www/html/wp-content/plugins/gutenberg/packages/style-engine/phpunit/class-wp-style-engine-test.php`
